File: src/int2char.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

unsigned short log10p1(long long int u)
{
	unsigned short exp = 1;
	while(u /= 10) exp ++;
	return exp;
}

int main(int argc, char* argv[])
{
	if(argc == 1)
	{
		printf("Usage: int2char [INTEGER]...\n"\
				"\tconvert INTEGER to UTF-8 - negatives (< 0) and larges (>= 2,097,152) will be ignored.\n");
		return 0;
	}
	int i, rpt = 1;
	while(rpt < argc)
	{
		long long int i = atoll(argv[rpt]);
		#ifdef VERIFY_OVERFLOW
		if(log10p1(i) > 17) fprintf(stderr, "EXCEPTION\n");
		else
		#endif
			if(i < 0) { fprintf(stderr, "Ignored negative number: %lld\n", i); }
		else if(i < 256) printf("%c", (int) i);
		else if(i < 2048)
		{
			char k[2] = {192 + (i >> 6), 128 + i % 64};
			printf("%c%c", k[0], k[1]);
		} else if(i < 65536)
		{	
			char k[3] = {
						224 + (i >> 12),
			   			128 + (i >> 6) % 64,
						128 + i % 64
			};
			printf("%c%c%c", k[0], k[1], k[2]);
		} else if(i < 2097152)
		{
			char k[4] = {
						240 + (i >> 18),
						128 + (i >> 12) % 64,
						128 + (i >> 6) % 64,
						128 + i % 64
			};
			printf("%c%c%c%c", k[0], k[1], k[2], k[3]);
		} else { fprintf(stderr, "Ignored large number: %lld\n", i); }
		rpt ++;
	}
	return 0;
}

File: src/int2char.rs
use std::{env, process::exit};


fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() == 1
    {
        println!("Usage: int2char [INTEGER]...\n\
				\tconvert INTEGER to UTF-8 - negatives (< 0) and larges (>= 2,097,152) will be ignored.\n");
        exit(0);
    }


    for element in &args[1..] {
        match string_to_natural(&element) {
            Some(t) => {
                if t < 256 {
                    print!("{}", t as u8 as char);
                } else if t < 2048 {
                    let arr = [(192 + (t >> 6)) as u8, (128 + t % 64) as u8,0,0];
                    print!("{}", String::from_utf8_lossy(&arr));
                } else if t < 65536 {
                    let arr = [(224 + (t >> 12)) as u8 , (128 + (t >> 6) % 64) as u8, (128 + t % 64) as u8,0];
                    print!("{}", String::from_utf8_lossy(&arr));
                } else if t < 2097152 {
                    let arr = [(240 + (t >> 18)) as u8 , (128 + (t >> 12) % 64) as u8, (128 + (t >> 6) % 64) as u8, (128 + t % 64) as u8];
                    print!("{}", String::from_utf8_lossy(&arr));
                } else {
                    eprintln!("Too large: {t}");
                }
                0
            },
            None => 0,
        };
    }

    exit(0);
}

fn exp10(v: u8) -> u32 {
    let mut c: u32 = 1;
    for _ in 1..v + 1 {
        c *= 10;
    }
    c
}

fn string_to_natural(str: &String) -> Option<u32> {
    let mut sum: u32 = 0;
    let mut count:u8 = 0;
    let str = str.trim_start_matches("0");
    for ch in str.bytes().rev() {
        sum +=  exp10(count) * (match ch {
            b'0' => 0,
            b'1' => 1,
            b'2' => 2,
            b'3' => 3,
            b'4' => 4,
            b'5' => 5,
            b'6' => 6,
            b'7' => 7,
            b'8' => 8,
            b'9' => 9,
            _ => { return None; }
        });
        count += 1;
    }
    Some(sum)
}
